airbyte-integrations/connectors/source-jira/source_jira/components/partition_routers.py

```diff
@@ -0,0 +1,104 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+from dataclasses import dataclass
+from typing import Any, Iterable, Mapping
+
+import dpath.util
+from airbyte_cdk.models import AirbyteMessage, SyncMode, Type
+from airbyte_cdk.sources.declarative.partition_routers.substream_partition_router import SubstreamPartitionRouter
+from airbyte_cdk.sources.declarative.types import Record, StreamSlice
+
+
+@dataclass
+class SubstreamPartitionRouterWithContext(SubstreamPartitionRouter):
+    """
+    It is impossible to pass additional data from the parent record to subsequent stream slices.
+    So, in this customization, we have prepared a small fix by setting the parent record data as an stream_slice.parent_record attribute
+    """
+
+    def stream_slices(self) -> Iterable[StreamSlice]:
+        if not self.parent_stream_configs:
+            yield from []
+        else:
+            for parent_stream_config in self.parent_stream_configs:
+                parent_stream = parent_stream_config.stream
+                parent_field = parent_stream_config.parent_key.eval(self.config)  # type: ignore # parent_key is always casted to an interpolated string
+                partition_field = parent_stream_config.partition_field.eval(self.config)  # type: ignore # partition_field is always casted to an interpolated string
+                for parent_stream_slice in parent_stream.stream_slices(
+                    sync_mode=SyncMode.full_refresh, cursor_field=None, stream_state=None
+                ):
+                    empty_parent_slice = True
+                    parent_partition = parent_stream_slice.partition if parent_stream_slice else {}
+
+                    for parent_record in parent_stream.read_records(
+                        sync_mode=SyncMode.full_refresh, cursor_field=None, stream_slice=parent_stream_slice, stream_state=None
+                    ):
+                        # Skip non-records (eg AirbyteLogMessage)
+                        if isinstance(parent_record, AirbyteMessage):
+                            if parent_record.type == Type.RECORD:
+                                parent_record = parent_record.record.data
+                            else:
+                                continue
+                        elif isinstance(parent_record, Record):
+                            parent_record = parent_record.data
+                        try:
+                            partition_value = dpath.util.get(parent_record, parent_field)
+                        except KeyError:
+                            pass
+                        else:
+                            empty_parent_slice = False
+                            stream_slice = StreamSlice(
+                                partition={partition_field: partition_value, "parent_slice": parent_partition}, cursor_slice={}
+                            )
+                            setattr(stream_slice, "parent_record", parent_record)
```

Comment:
why do we need the full record as opposed to the just the parent record id in the slice?

Reply:
We often require certain data from the parent record to be transferred to the child record. For example, in the HubSpot connector, there are streams where parent streams have composite primary keys, and we need to include all the fields that constitute the primary key into the child record. Here we pass some information to the child record as well, and of course, we don't need all the data. I will optimize this as part of future refactoring.